web/text_layer_builder.js

```diff
@@ -181,44 +184,116 @@ class TextLayerBuilder {
+    let prevRange = null;
+
     div.addEventListener("mousedown", evt => {
-      const end = div.querySelector(".endOfContent");
-      if (!end) {
-        return;
-      }
-      if (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) {
-        // On non-Firefox browsers, the selection will feel better if the height
-        // of the `endOfContent` div is adjusted to start at mouse click
-        // location. This avoids flickering when the selection moves up.
-        // However it does not work when selection is started on empty space.
-        let adjustTop = evt.target !== div;
-        if (typeof PDFJSDev === "undefined" || PDFJSDev.test("GENERIC")) {
-          adjustTop &&=
-            getComputedStyle(end).getPropertyValue("-moz-user-select") !==
-            "none";
-        }
-        if (adjustTop) {
-          const divBounds = div.getBoundingClientRect();
-          const r = Math.max(0, (evt.pageY - divBounds.top) / divBounds.height);
-          end.style.top = (r * 100).toFixed(2) + "%";
-        }
-      }
+      selectingThroughMouse = true;
       end.classList.add("active");
     });
 
-    div.addEventListener("mouseup", () => {
-      const end = div.querySelector(".endOfContent");
-      if (!end) {
-        return;
-      }
+    const reset = () => {
       if (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) {
-        end.style.top = "";
+        div.append(end);
+        end.style.width = "";
+        end.style.height = "";
       }
       end.classList.remove("active");
-    });
+      selectingThroughMouse = false;
+      prevRange = null;
+    };
+    div.addEventListener("pointerup", reset);
+
+    // On touchscreen devices, when expaning/reducing a selection through the
+    // 'bubbles' that mark the selection end/start the browser will not fire
+    // any pointer event, because the user is interacting with browser-specific
+    // UI and not with the webpage itself.
+    // To workaround this, we listen to the 'selectionchange' event and check if
+    // the selection includes elements from this page. This has two drawbacks:
+    //  - We can only detect changes once they start happening. If the first
+    //    movement that the user does is to move one of the selection bubbles
+    //    to a blank part of the page, the selection will flicker before
+    //    that we can apply the fix.
+    //  - We cannot detect when the user stops interacting with the selection,
+    //    because there is no difference between "the user is still holding
+    //    their finger on the selection 'bubble'" and "the user is lifter their
+    //    finger and then pressed the 'bubble' again". Thus, the .endOfContent
+    //    element will remain active until when the user interacts with the
+    //    page again, triggering a pointerup event.
+    this.#selectionChangeAbortController?.abort();
+    this.#selectionChangeAbortController = new AbortController();
+    document.addEventListener(
```

Comment:
Every text layer will add its listener for the `selectionchange` event, would it be possible to avoid that in having only one listener ?

Reply:
And when the text layer is destroyed (for example when scrolling), the listener isn't removed or did I miss something ?

Reply:
I could use a WeakMap to go from `.textLayer` to this class, and then in a global event listener I dispatch to the right handler. Let me try.

Reply:
Ah yes correct: I looked for removeEventListener...

Reply:
By trying to move to a single global `selectionchange` event listener I realized that:
- multi-page selection are possible (a single selection can span across multiple `.textLayer`s)
- multi-page selection currently flicker everywhere (Firefox&Chrome, mobile&desktop)